### PR TITLE
8.0 update SignalR config: withServerTimeout, withKeepAliveInterval

### LINKED ISSUE #29717: 8.0: SignalR Config: Update for ServerTimeout and KeepAliveInterval on HubConnectionBuilder
In the 8.0 version of the client section [Configure timeout and keep-alive options](https://learn.microsoft.com/en-us/aspnet/core/signalr/configuration?view=aspnetcore-8.0&tabs=dotnet#configure-timeout-and-keep-alive-options-6) should be updated to include the new 8.0 approach for setting ServerTimeout and KeepAliveInterval directly on [HubConnectionBuilder](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.signalr.client.hubconnectionbuilder?view=aspnetcore-8.0) as demonstrated in the [What's new in 8.0 topic](https://learn.microsoft.com/en-us/aspnet/core/release-notes/aspnetcore-8.0?view=aspnetcore-7.0#new-approach-to-set-the-server-timeout-and-keep-alive-interval).  

Also see related PR focused on Blazor: Server timeout and Keep Alive interval config PR #29297 and current published versions of those topics with minor updates.

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: de8cdc99-48bb-3266-05bb-eda98f59f2e2
* Version Independent ID: 66d252c3-6300-7e28-9aa0-ef2ff66e7a76
* Content: [ASP.NET Core SignalR configuration](https://learn.microsoft.com/en-us/aspnet/core/signalr/configuration?view=aspnetcore-8.0&tabs=javascript#configure-timeout-and-keep-alive-options-6)
* Content Source: [aspnetcore/signalr/configuration.md](https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/signalr/configuration.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-signalr**
* GitHub Login: @bradygaster
* Microsoft Alias: **bradyg**

---
[Associated WorkItem - 123574](https://dev.azure.com/msft-skilling/Content/_workitems/edit/123574)

### PR DESCRIPTION
**Pull request by wadepickett**
[Internal Review:  Version 8.0 Client Timeout and keep-alive Options section: .NET & JavaScript tabs](https://review.learn.microsoft.com/en-us/aspnet/core/signalr/configuration?view=aspnetcore-8.0&branch=pr-en-us-29761&tabs=dotnet#configure-timeout-and-keep-alive-options)

Fixes #29717 
8.0 updates of SingnalR configuration for withServerTimeout and with withKeepAliveInterval.

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [aspnetcore/signalr/configuration.md](https://github.com/dotnet/AspNetCore.Docs/blob/c7c8b303fa55f973ec49d331a20634a618b161c1/aspnetcore/signalr/configuration.md) | [ASP.NET Core SignalR configuration](https://review.learn.microsoft.com/en-us/aspnet/core/signalr/configuration?branch=pr-en-us-29761) |

### CODE REVIEW

File: aspnetcore/signalr/configuration.md [hunk 1]
```diff
@@ -278,18 +278,46 @@ Additional options for configuring timeout and keep-alive behavior are available
 
 | Option | Default value | Description |
 | ------ | ------------- | ----------- |
-| `ServerTimeout` | 30 seconds (30,000 milliseconds) | Timeout for server activity. If the server hasn't sent a message in this interval, the client considers the server disconnected and triggers the `Closed` event (`onclose` in JavaScript). This value must be large enough for a ping message to be sent from the server **and** received by the client within the timeout interval. The recommended value is a number at least double the server's `KeepAliveInterval` value to allow time for pings to arrive. |
+| `withServerTimeout` | 30 seconds (30,000 milliseconds) | Timeout for server activity in milliseconds. If the server hasn't sent a message in this interval, the client considers the server disconnected and triggers the `Closed` event (`onclose` in JavaScript). This value must be large enough for a ping message to be sent from the server **and** received by the client within the timeout interval. The recommended value is a number at least double the server's Keep-Alive interval (`withKeepAliveInterval`) value to allow time for pings to arrive. |
```
Comment (tdykstra) [inline aspnetcore/signalr/configuration.md:281]: I think the text version of `KeepAliveInterval` should not be capitalized -- keep-alive interval, like you do with server timeout.

File: aspnetcore/signalr/configuration.md [hunk 2]
```diff
@@ -278,18 +278,46 @@ Additional options for configuring timeout and keep-alive behavior are available
...
 
 In the .NET Client, timeout values are specified as `TimeSpan` values.
 
+In ASP.NET Core 8.0 or later <xref:Microsoft.AspNetCore.SignalR.Client.HubConnection.ServerTimeout> (default: 30 seconds) and <xref:Microsoft.AspNetCore.SignalR.Client.HubConnection.KeepAliveInterval> (default: 15 seconds) can be set directly on <xref:Microsoft.AspNetCore.SignalR.Client.HubConnectionBuilder>.
+
+The following example shows assigning values that are double the default values:
```
Comment (tdykstra) [inline aspnetcore/signalr/configuration.md:289]:
```suggestion
The following example shows values that are double the default values:
```

File: aspnetcore/signalr/configuration.md [hunk 3]
```diff
@@ -278,18 +278,46 @@ Additional options for configuring timeout and keep-alive behavior are available
...
+
+In ASP.NET Core 8.0 or later <xref:Microsoft.AspNetCore.SignalR.Client.HubConnection.ServerTimeout> (default: 30 seconds) and <xref:Microsoft.AspNetCore.SignalR.Client.HubConnection.KeepAliveInterval> (default: 15 seconds) can be set directly on <xref:Microsoft.AspNetCore.SignalR.Client.HubConnectionBuilder>.
+
+The following example shows assigning values that are double the default values:
```
Comment (tdykstra) [inline aspnetcore/signalr/configuration.md:312]:
```suggestion
The following example shows values that are double the default values:
```

File: aspnetcore/signalr/configuration.md [hunk 4]
```diff
@@ -278,18 +278,46 @@ Additional options for configuring timeout and keep-alive behavior are available
...
+| `withKeepAliveIntervalInMilliseconds` | 15 seconds (15,000 milliseconds) | Determines the interval at which the client sends ping messages. This setting allows the server to detect hard disconnects, such as when a client unplugs their computer from the network. Sending any message from the client resets the timer to the start of the interval. If the client hasn't sent a message in the `ClientTimeoutInterval` set on the server, the server considers the client disconnected. The ping occurs at most as often as the server pings. If the server pings every five seconds, assigning a value lower than `5000` (5 seconds) pings every five seconds. The default value is 15 seconds. The Keep-Alive interval should be less than or equal to half the value assigned to the server timeout (`withServerTimeout`). |
+
+In ASP.NET Core 8.0 or later <xref:Microsoft.AspNetCore.SignalR.Client.HubConnection.ServerTimeout> (default: 30 seconds) and <xref:Microsoft.AspNetCore.SignalR.Client.HubConnection.KeepAliveInterval> (default: 15 seconds) can be set directly on <xref:Microsoft.AspNetCore.SignalR.Client.HubConnectionBuilder>.
```
Comment (tdykstra) [inline aspnetcore/signalr/configuration.md:310]:
```suggestion
In ASP.NET Core 8.0 or later, <xref:Microsoft.AspNetCore.SignalR.Client.HubConnection.ServerTimeout> (default: 30 seconds) and <xref:Microsoft.AspNetCore.SignalR.Client.HubConnection.KeepAliveInterval> (default: 15 seconds) can be set directly on <xref:Microsoft.AspNetCore.SignalR.Client.HubConnectionBuilder>.
```

File: aspnetcore/signalr/configuration.md (hunk 1 above)
Comment (tdykstra) [inline aspnetcore/signalr/configuration.md:281]: If this is the .NET client tab, and .NET client uses TimeSpans for timeout durations, is it correct to say the timeout is "in milliseconds"?

**Review by tdykstra — APPROVED** [review]
LGTM, aside from a few nits for you to consider.

File: aspnetcore/signalr/configuration.md [hunk 5]
```diff
@@ -278,18 +278,46 @@ Additional options for configuring timeout and keep-alive behavior are available
 
 | Option | Default value | Description |
 | ------ | ------------- | ----------- |
-| `ServerTimeout` | 30 seconds (30,000 milliseconds) | Timeout for server activity. If the server hasn't sent a message in this interval, the client considers the server disconnected and triggers the `Closed` event (`onclose` in JavaScript). This value must be large enough for a ping message to be sent from the server **and** received by the client within the timeout interval. The recommended value is a number at least double the server's `KeepAliveInterval` value to allow time for pings to arrive. |
```
Comment (BrennanConroy) [inline aspnetcore/signalr/configuration.md:281]: 7 lines above this:
> Additional options for configuring timeout and keep-alive behavior are available on the `HubConnection` object itself:

The methods are on `HubConnectionBuilder`, so this would need to be reworked if you go this approach.

File: aspnetcore/signalr/configuration.md [hunk 6]
```diff
@@ -278,18 +278,46 @@ Additional options for configuring timeout and keep-alive behavior are available
...
-| `KeepAliveInterval` | 15 seconds | Determines the interval at which the client sends ping messages. Sending any message from the client resets the timer to the start of the interval. If the client hasn't sent a message in the `ClientTimeoutInterval` set on the server, the server considers the client disconnected. |
+| `withKeepAliveInterval` | 15 seconds | Determines the interval in milliseconds at which the client sends ping messages. This setting allows the server to detect hard disconnects, such as when a client unplugs their computer from the network. Sending any message from the client resets the timer to the start of the interval. If the client hasn't sent a message in the `ClientTimeoutInterval` set on the server, the server considers the client disconnected. The ping occurs at most as often as the server pings. If the server pings every five seconds, assigning a value lower than `5000` (5 seconds) pings every five seconds. The default value is 15 seconds. The Keep-Alive interval should be less than or equal to half the value assigned to the server timeout (`withServerTimeout`). |
```
Comment (BrennanConroy) [inline aspnetcore/signalr/configuration.md:283]: Methods are Pascal cased in .NET

File: aspnetcore/signalr/configuration.md [hunk 7]
```diff
@@ -278,18 +278,46 @@ Additional options for configuring timeout and keep-alive behavior are available
...
 
 In the .NET Client, timeout values are specified as `TimeSpan` values.
 
+In ASP.NET Core 8.0 or later <xref:Microsoft.AspNetCore.SignalR.Client.HubConnection.ServerTimeout> (default: 30 seconds) and <xref:Microsoft.AspNetCore.SignalR.Client.HubConnection.KeepAliveInterval> (default: 15 seconds) can be set directly on <xref:Microsoft.AspNetCore.SignalR.Client.HubConnectionBuilder>.
```
Comment (BrennanConroy) [inline aspnetcore/signalr/configuration.md:287]: Why is this needed if you are getting rid of `ServerTimeout` and `KeepAliveInterval` from the docs?

Comment (wadepickett) [reply aspnetcore/signalr/configuration.md:287]: Thanks, I can remove it.  I was a bit worried that developers already using SignalR for previous versions might be wondering what was happening there.

File: aspnetcore/signalr/configuration.md (hunk 5 above)
Comment (wadepickett) [reply aspnetcore/signalr/configuration.md:281]: Resolved in latest commit.  Thanks very much Brennan for the great help.

File: aspnetcore/signalr/configuration.md [hunk 8]
```diff
@@ -272,24 +272,50 @@ HubConnection hubConnection = HubConnectionBuilder.create("https://example.com/c
...
+| `WithServerTimeout` | 30 seconds (30,000 milliseconds) | Timeout for server activity. If the server hasn't sent a message in this interval, the client considers the server disconnected and triggers the `Closed` event (`onclose` in JavaScript). This value must be large enough for a ping message to be sent from the server **and** received by the client within the timeout interval. The recommended value is a number at least double the server's keep-alive interval (`WithKeepAliveInterval`) value to allow time for pings to arrive. |
 | `HandshakeTimeout` | 15 seconds | Timeout for initial server handshake. If the server doesn't send a handshake response in this interval, the client cancels the handshake and triggers the `Closed` event (`onclose` in JavaScript). This is an advanced setting that should only be modified if handshake timeout errors are occurring due to severe network latency. For more detail on the handshake process, see the [SignalR Hub Protocol Specification](https://github.com/aspnet/SignalR/blob/master/specs/HubProtocol.md). |
```
Comment (BrennanConroy) [inline aspnetcore/signalr/configuration.md:282]: This property is still only on `HubConnection`

File: aspnetcore/signalr/configuration.md [hunk 9]
```diff
@@ -272,24 +272,50 @@ HubConnection hubConnection = HubConnectionBuilder.create("https://example.com/c
...
+| `withKeepAliveIntervalInMilliseconds` | 15 seconds (15,000 milliseconds) | Determines the interval at which the client sends ping messages. This setting allows the server to detect hard disconnects, such as when a client unplugs their computer from the network. Sending any message from the client resets the timer to the start of the interval. If the client hasn't sent a message in the `ClientTimeoutInterval` set on the server, the server considers the client disconnected. The ping occurs at most as often as the server pings. If the server pings every five seconds, assigning a value lower than `5000` (5 seconds) pings every five seconds. The default value is 15 seconds. The keep-alive interval should be less than or equal to half the value assigned to the server timeout (`withServerTimeout`). |
+
+In ASP.NET Core 8.0 or later, <xref:Microsoft.AspNetCore.SignalR.Client.HubConnection.ServerTimeout> (default: 30 seconds) and <xref:Microsoft.AspNetCore.SignalR.Client.HubConnection.KeepAliveInterval> (default: 15 seconds) can be set directly on <xref:Microsoft.AspNetCore.SignalR.Client.HubConnectionBuilder>.
```
Comment (BrennanConroy) [inline aspnetcore/signalr/configuration.md:308]: This is still here?

File: aspnetcore/signalr/configuration.md (hunk 8 above)
Comment (wadepickett) [reply aspnetcore/signalr/configuration.md:282]: Thanks, I will move the detail of `HubConnection` object and `HubConnectionBuilder` to each property.

Comment (wadepickett) [reply aspnetcore/signalr/configuration.md]: Fixed in commit e99b087

File: aspnetcore/signalr/configuration.md (hunk 9 above)
Comment (wadepickett) [reply aspnetcore/signalr/configuration.md]: Fixed now in commit [c7c8b30](https://github.com/dotnet/AspNetCore.Docs/pull/29761/commits/c7c8b303fa55f973ec49d331a20634a618b161c1)

**Comment by wadepickett** [comment]
Thanks for the great feedback.  Sorry Brennan I had messed up the fix originally.  I will go ahead and merge.

File: aspnetcore/signalr/configuration.md [hunk 10]
```diff
@@ -272,24 +272,48 @@ HubConnection hubConnection = HubConnectionBuilder.create("https://example.com/c
...
+| `WithKeepAliveInterval` | 15 seconds | Determines the interval at which the client sends ping messages and is set directly on <xref:Microsoft.AspNetCore.SignalR.Client.HubConnectionBuilder>. This setting allows the server to detect hard disconnects, such as when a client unplugs their computer from the network. Sending any message from the client resets the timer to the start of the interval. If the client hasn't sent a message in the `ClientTimeoutInterval` set on the server, the server considers the client disconnected. The ping occurs at most as often as the server pings. If the server pings every five seconds, assigning a value lower than `5000` (5 seconds) pings every five seconds. The default value is 15 seconds. The keep-alive interval should be less than or equal to half the value assigned to the server timeout (`WithServerTimeout`). |
```
Comment (BrennanConroy) [inline aspnetcore/signalr/configuration.md:283]: Remove the extra content that was added here. It only applies to the javascript section.

Comment (wadepickett) [reply aspnetcore/signalr/configuration.md:283]: Thanks @BrennanConroy, I will fix immediately in a new PR, but which extra content are you referring to?   

Keep the information on `WithKeepAlive`Interval but remove "The ping occurs at most as often as the server pings. If the server pings every five seconds.."?   

Or do you mean back out of everything added to this .NET tab section including `WithServerTimeout` and `WithKeepAliveInterval`?

Comment (BrennanConroy) [reply aspnetcore/signalr/configuration.md:283]: > "The ping occurs at most as often as the server pings. If the server pings every five seconds.."?

That part.

Comment (wadepickett) [reply aspnetcore/signalr/configuration.md:283]: Thanks!!!  Creating a PR to remove that right now since this one already published (but too soon).

Comment (wadepickett) [reply aspnetcore/signalr/configuration.md:283]: Fixed in PR #29814 and ready for review.
